### ISSUE TITLE
ZFS diff causes hard lockups when file has invalid parent object

### ISSUE THREAD

**Issue by gbooker** [zfsonlinux/zfs#2597]
This may be the same as #2139 but my dmesg looks different so I'm posting in another bug just in case.  Also, my FS did require #1927 before it was able to start doing any diffs.

If I run a zfs diff between two snapshots on my FS, it will start performing the diff and then hard lock in the same place.  It seems that when it locks, the entire storage system is inaccessible.  I described much of the process in the mailing list: https://groups.google.com/a/zfsonlinux.org/forum/#!topic/zfs-discuss/qKH9-uw8eTk  I'll repeat the most important aspects here as well as additional findings.

I run "zfs diff zroot/media@20121010 zroot/media@20121201" and it starts the process of the diff.  Part of the way through the diff, the process hangs.  Once, and only once, I was able to get the output of dmesg while it was hung.  Most of the time I can't do anything at all during the hang.
https://gist.github.com/gbooker/9efd05157114b5e8365d

The disk mentioned in the dmesg output, ata8.00/sdd, is the boot drive, a 60GB Intel SSD.  As a test, I boot from an Xubuntu USB install with this drive disconnected.  When I ran the diff, it locked again, only this time it complained about ata5, which is the BD drive (which had no media in the drive and hadn't been used).

Since my motherboard had 6 SATA ports on it, I moved all drives to those ports and removed my 4 port SATA card.  Repeated the test with the USB drive and it locked again (saw no information about which drive it complained about this time).

Finally, I fashioned a USB drive with FreeBSD 10 on it (with all the hardware connected again).  Ran the same diff and it finished normally.  It ended the diff with the line:
Unable to determine path or stats for object 14941 in zroot/media@20121010: Invalid argument

I ran zdb on this object ID in Linux, and it CPU spun before the line printing out the path.  The machine was not locked and top reported zdb at 100% CPU usage.  Ctrl-C did kill the zdb process.  I was able to run zdb in FreeBSD and it reported:
https://gist.github.com/gbooker/901249aad60bfba08c0c

The parent object, 14940, is another file.  Google yielded another who had the same problem, which appears to not actually be problem:
http://lists.freebsd.org/pipermail/freebsd-fs/2012-November/015654.html
which was then followed up by:
https://www.mail-archive.com/zfs-discuss@opensolaris.org/msg50086.html
The latter thread describes a process in which this can occur.  While I don't use hard links often, I have used it from time to time on this FS.

If I understand this correctly, it appears that using hard links results in a possibly invalid parent object id.  When zfs diff encounters such an id, in my case the parent id was a different file object, it hard locks the machine.  So, zfs diff should be more tolerant of this situation and print the path as unknown or try to ascertain the real path or paths.

**Comment by Ringdingcoder** [comment]
It turns out that the previous zdb output was misleading because I used it on the wrong file system – the root of the pool instead of the child on which I ran the diff :)

Having read the opensolaris thread from 2012 linked above, it seems like a wrong parent pointer is indeed causing this. Object 21 is a file object created in 2009, shortly after I originally created the pool. Its parent points to an object from 2013. It is not supposed to be that way, I guess.

I’ll upgrade my machine to kernel 4.2.0 and zol 0.6.5 and see what happens.
